docs/proposals/performance-related-monitoring.md

```diff
@@ -0,0 +1,138 @@
+
+In late September we encountered a strange problem (https://github.com/kubernetes/kubernetes/issues/14554): we observed an increased observed latencies in small clusters (few
+Nodes). It turned out that it’s caused by an added latency between PodRunning and PodReady phases. This was not a real regression, but our tests thought it were, which shows
+how careful we need to be.
+
+### Huge number of handshakes slows down API server
+
+It was a long standing issue for performance and is/was an important bottleneck for scalability (https://github.com/kubernetes/kubernetes/issues/13671). The bug directly
+causing this problem was incorrect (from the golangs standpoint) handling of TCP connections. Secondary issue was that elliptic curve encryption (only one available in go 1.4)
+is unbelievably slow.
+
+## Proposed metrics/statistics to gather/compute to avoid problems
+
+### Logging monitoring
+
+Log spam is a serious problem and we need to keep it under control. Simplest way to check for regressions, suggested by @bredanburns, is to compute the rate in which log files
+grow in e2e tests.
+
+Basic ideas:
+- log generation rate (B/s)
+
+### REST call monitoring
+
+We do measure REST call duration in the Density test, but we need an API server monitoring as well, to avoid false failures caused e.g. by the network traffic. We already have
+some metrics in place (https://github.com/kubernetes/kubernetes/blob/master/pkg/apiserver/metrics/metrics.go), but we need to revisit the list and add some more.
+
+Basic ideas:
+- number of calls per verb, client, resource type
+- latency distribution per verb, client, resource type
+- number of calls that was rejected per client, resource type and reason (invalid version number, already at maximum number of requests in flight)
+
+### Rate limit monitoring
+
+Reverse of REST call monitoring done in the API server. We need to know when a given component increases a pressure it puts on the API server. As a proxy for number of
+requests sent we can track how saturated are rate limiters. This has additional advantage of giving us data needed to fine-tune rate limiter constants.
+
+Because we have rate limitting on both ends (client and API server) we should monitor number of inflight requests in API server and how it relates to `max-requests-inflight`.
+
+Basic ideas:
+- percentage of used non-burst limit,
+- amount of time in last hour with depleted burst tokens,
+- number of inflight requests in API server.
+
+### Network connection monitoring
+
+During development we observed incorrect use/reuse of HTTP connections multiple times already. We should at least monitor number of created connections.
+
+### ETCD monitoring
+
+@xiang-90 and @hongchaodeng - you probably have way more experience on what'd be good to look at from the ETCD perspective.
+
+Basic ideas:
+- ETCD memory footprint
+- number of objects per kind
+- read/write latencies per kind
+- number of requests from the API server
+
```

Comment:
if we could enable a debug flag that gave read+write counts per key that would be useful as well.